packages/core/src/parser.ts

```diff
@@ -0,0 +1,155 @@
+    }
+
+    const pathParams = Object.create(null);
+    for (const ix in endPoint._pathRegexp.keys) {
+      const key = endPoint._pathRegexp.keys[ix];
+      const matchIndex = +ix + 1;
+      pathParams[key.name] = match[matchIndex];
+    }
+
+    const operationName = endPoint._spec['x-operation-name'];
+    const body = loadRequestBodyIfNeeded(endPoint._spec, request)
+    return buildOperationArguments(endPoint._spec, parsedRequest, pathParams, body);
+  }
+  
+
+interface ParsedRequest extends Request {
+  // see http://expressjs.com/en/4x/api.html#req.path
+  path: string;
+  // see http://expressjs.com/en/4x/api.html#req.query
+  query: { [key: string]: string };
+  // see https://github.com/DefinitelyTyped/DefinitelyTyped/issues/15808
+  url: string;
+  pathname: string;
+  method: string;
+}
+
+function loadRequestBodyIfNeeded(operationSpec: OperationObject, request: Request): Promise<MaybeBody> {
+  if (!hasArgumentsFromBody(operationSpec))
+    return Promise.resolve();
+
+  const contentType = request.headers['content-type'];
+  if (contentType && !/json/.test(contentType)) {
+    const err = createHttpError(415, `Content-type ${contentType} is not supported.`);
+    return Promise.reject(err);
+  }
+
+  return parseJsonBody(request).catch((err: HttpError) => {
+    err.statusCode = 400;
+    return Promise.reject(err);
+  });
+}
+
+function hasArgumentsFromBody(operationSpec: OperationObject): boolean {
+  if (!operationSpec.parameters || !operationSpec.parameters.length)
+   return false;
+
+  for (const paramSpec of operationSpec.parameters) {
+    if ('$ref' in paramSpec) continue;
+    const source = (paramSpec as ParameterObject).in;
+    if (source === 'formData' || source === 'body')
+     return true;
+  }
+  return false;
+}
+
+function buildOperationArguments(operationSpec: OperationObject, request: ParsedRequest,
+    pathParams: PathParameterValues, body?: MaybeBody): OperationArgs {
+  const args: OperationArgs = [];
+
+  for (const paramSpec of operationSpec.parameters || []) {
+    if ('$ref' in paramSpec) {
+      // TODO(bajtos) implement $ref parameters
+      throw new Error('$ref parameters are not supported yet.');
+    }
+    const spec = paramSpec as ParameterObject;
+    switch (spec.in) {
+      case 'query':
+        args.push(request.query[spec.name]);
+        break;
+      case 'path':
+        args.push(pathParams[spec.name]);
+        break;
+      case 'header':
+        args.push(request.headers[spec.name.toLowerCase()]);
+        break;
+      case 'formData':
+        args.push(body ? body[spec.name] : undefined);
+        break;
+      case 'body':
+        args.push(body);
+        break;
+      default:
+        throw createHttpError(501, 'Parameters with "in: ' + spec.in + '" are not supported yet.');
+    }
+  }
+  return args;
+}
+
+interface HttpError extends Error {
+  statusCode?: number;
+  status?: number;
+}
+
+function createHttpError(statusCode: number, message: string) {
```

Comment:
As we have discussed elsewhere (sorry, don't have a link), creating an HTTP error object is so frequent case that LoopBack should provide a public API for that. One option is to leverage [http-errors](https://www.npmjs.com/package/http-errors) module, as we are already doing in CodeHub's [UserController](https://github.com/strongloop/loopback-next/blob/57a6868632638b42e99e7bf44882d83dd570e939/packages/example-codehub/src/controllers/UserController.ts).

Having said that, it's probably best to leave this improvement out of this initial pull request.